lib/routes/account.js

```diff
@@ -1021,50 +1048,78 @@ module.exports = function (
-        db.account(Buffer(uid, 'hex'))
-          .then(
-            function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', { email: account.email, uid: account.uid, locale: account.locale })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid)
+        if(code.length === 32) {
+          // Hex code length of 32 signifies a tokenId, perform login token verification
+          log.begin('Account.RecoveryLoginVerify', request);
+          db.verifyToken(code, {uid: uid})
+            .then(function () {
+            })
+            .done(
+              function () {
+                reply({})
+              },
+              reply
+            )
+        } else {
+          log.begin('Account.RecoveryEmailVerify', request)
+          db.account(Buffer(uid, 'hex'))
+            .then(
+              function (account) {
+                // If the account is already verified, they may be e.g.
+                // clicking a stale link.  Silently succeed.
+                if (account.emailVerified) {
+                  log.increment('account.already_verified')
+                  return true
+                }
+                if (!butil.buffersAreEqual(code, account.emailCode)) {
+                  throw error.invalidVerificationCode()
+                }
+                log.timing('account.verified', Date.now() - account.createdAt)
+                log.event('verified', {email: account.email, uid: account.uid, locale: account.locale})
+                log.increment('account.verified')
 
-              return db.verifyEmail(account)
-                .then(mailer.sendPostVerifyEmail.bind(
-                    mailer,
-                    account.email,
-                    {
-                      acceptLanguage: request.app.acceptLanguage
-                    }
-                  )
-                )
-            }
-          )
-          .done(
-            function () {
-              reply({})
-            },
-            reply
-          )
+                // send a push notification to all devices that the account changed
+                push.notifyUpdate(uid)
+
+                return db.sessions(account.uid)
+                  .then(function(sessions) {
+                    // TODO How to know which session is associated with the email verification?
+                    // Assuming last created session?
```

Comment:
> var code = crypto.randomBytes(16)
> db.createAccount({..., emailCode: code, ...})
> db.createSessionToken({..., tokenVerificationId: code, ... })
> db.createKeyFetchToken({..., tokenVerificationId: code, ... })

@vbudhram's explanation to me just made the penny drop as @philbooth says. Passing in the `emailCode` to the account on creation so that it can be re-used across multiple unverified account sign-ins also breaks the sandbox.

Here's the scenario @vbudhram  and I spoke about:

1. A user with less than stellar password practices (they re-use email/password combinations) signs up for Sync, but does not verify.
1. A bad actor finds the user's email and password from a hacked site. They sign in as the user and wait. The attacker is given the same emailCode as the user. A new email is sent to the user to verify their sign in, the user realizes they never verified and clicks the verification link. 
1. The attacker is now fully verified as the user. Wah wah.